### PR TITLE
archivebox: refactor and fix

### PR DESCRIPTION
**Pull request by Jonasmadsen**
###### Motivation for this change
Archivebox didn't build due to conflicting dependency on django.
Possibly related to Issue https://github.com/NixOS/nixpkgs/issues/154950 that was closed by PR https://github.com/NixOS/nixpkgs/pull/154153 

###### Things done
Resolves conflict by overriding django dependency for all dependencies.
Updated commit message and using packageOverrides as requested by reviewers on previous now closed PR https://github.com/NixOS/nixpkgs/pull/156520

Basic functionality tested and working.
Nothing else should depend on archivebox.

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))

- [ ] Tested, as applicable:
  - [X] NixOS
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [X] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/applications/misc/archivebox/default.nix [hunk 1]
```diff
@@ -1,27 +1,22 @@
...
-      sha256 = "sha256-Ms55Lum2oMu+w0ASPiKayfdl3/jCpK6SR6FLK6OjZac=";
+  py = python3.override {
+    packageOverrides = self: super: {
+      django = super.django_3.overridePythonAttrs (old: rec {
+        pname = "Django";
+        version = "3.1.7";
+        src = fetchPypi {
+          inherit pname version;
+          sha256 = "sha256-Ms55Lum2oMu+w0ASPiKayfdl3/jCpK6SR6FLK6OjZac=";
+        };
```
Comment (dotlambda) [inline pkgs/applications/misc/archivebox/default.nix:16]:
```suggestion
        src = old.src.override {
          inherit version;
          sha256 = "sha256-Ms55Lum2oMu+w0ASPiKayfdl3/jCpK6SR6FLK6OjZac=";
        };
```

File: pkgs/applications/misc/archivebox/default.nix [hunk 2]
```diff
@@ -1,27 +1,22 @@
...
-  django_3' = django_3.overridePythonAttrs (old: rec {
-    pname = "Django";
-    version = "3.1.7";
-    src = fetchPypi {
-      inherit pname version;
-      sha256 = "sha256-Ms55Lum2oMu+w0ASPiKayfdl3/jCpK6SR6FLK6OjZac=";
+  py = python3.override {
+    packageOverrides = self: super: {
+      django = super.django_3.overridePythonAttrs (old: rec {
+        pname = "Django";
```
Comment (dotlambda) [inline pkgs/applications/misc/archivebox/default.nix:11]:
```suggestion
        pname = "django";
```
https://github.com/NixOS/nixpkgs/pull/158710

File: pkgs/applications/misc/archivebox/default.nix [hunk 3]
```diff
@@ -1,27 +1,22 @@
 { lib
 , buildPythonApplication
+, python3
 , fetchPypi
```
Comment (dotlambda) [inline pkgs/applications/misc/archivebox/default.nix:4]:
```suggestion
, python3
```

File: pkgs/applications/misc/archivebox/default.nix [hunk 4]
```diff
@@ -1,27 +1,22 @@
...
+          inherit pname version;
+          sha256 = "sha256-Ms55Lum2oMu+w0ASPiKayfdl3/jCpK6SR6FLK6OjZac=";
+        };
+      });
     };
-  });
+  };
 in
 
 buildPythonApplication rec {
```
Comment (dotlambda) [inline pkgs/applications/misc/archivebox/default.nix:22]:
```suggestion
python.pkgs.buildPythonApplication rec {
```

File: pkgs/applications/misc/archivebox/default.nix [hunk 5]
```diff
@@ -1,27 +1,22 @@
...
 }:
 
 let
-  django_3' = django_3.overridePythonAttrs (old: rec {
-    pname = "Django";
-    version = "3.1.7";
-    src = fetchPypi {
-      inherit pname version;
-      sha256 = "sha256-Ms55Lum2oMu+w0ASPiKayfdl3/jCpK6SR6FLK6OjZac=";
+  py = python3.override {
```
Comment (dotlambda) [inline pkgs/applications/misc/archivebox/default.nix:8]:
```suggestion
  python = python3.override {
```

File: pkgs/applications/misc/archivebox/default.nix [hunk 6]
```diff
@@ -33,10 +28,10 @@ buildPythonApplication rec {
     sha256 = "sha256-zHty7lTra6yab9d0q3EqsPG3F+lrnZL6PjQAbL1A2NY=";
   };
 
-  propagatedBuildInputs = [
+  propagatedBuildInputs = with py.pkgs; [
```
Comment (dotlambda) [inline pkgs/applications/misc/archivebox/default.nix:31]:
```suggestion
  propagatedBuildInputs = with python.pkgs; [
```

**Review by dotlambda — CHANGES_REQUESTED** [review]
Please don't use `python3Packages.callPackage` in `all-packages.nix`.

**Comment by Jonasmadsen** [comment]
@dotlambda accepted your changes

File: pkgs/applications/misc/archivebox/default.nix [hunk 7]
```diff
@@ -1,42 +1,35 @@
...
+  };
 in
 
-buildPythonApplication rec {
+python.pkgs.buildPythonApplication rec {
   pname = "archivebox";
   version = "0.6.2";
 
-  src = fetchPypi {
+  src = python3.pkgs.fetchPypi {
```
Comment (dotlambda) [inline pkgs/applications/misc/archivebox/default.nix:24]:
```suggestion
  src = python.pkgs.fetchPypi {
```

**Comment by siraben** [comment]
Result of `nixpkgs-review pr 158704` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li>archivebox</li>
  </ul>
</details>

**Comment by dotlambda** [comment]
> 1 package failed to build

@siraben Was that due to ipython, like for me on aarch64-linux?

**Comment by siraben** [comment]
@dotlambda yes, that's the cause of the failure.

**Pull request merged**